File: web/app/services/gpts.ts
import { Gpts } from "@/app/types/gpts";
import fs from "fs";

export const getGptsFromFile = async (): Promise<Gpts[]> => {
  try {
    const dataFile = process.env.GPTS_DATA_FILE;
    if (!dataFile) {
      return [];
    }

    const data = fs.readFileSync(dataFile, "utf8");
    const jsonData = JSON.parse(data);

    let gpts: Gpts[] = [];
    jsonData.map((v: any) => {
      gpts.push({
        uuid: v["data"]["gizmo"]["id"],
        org_id: v["data"]["gizmo"]["organization_id"],
        name: v["data"]["gizmo"]["display"]["name"],
        description: v["data"]["gizmo"]["display"]["description"],
        avatar_url: v["data"]["gizmo"]["display"]["profile_picture_url"],
        short_url: v["data"]["gizmo"]["short_url"],
        author_id: v["data"]["gizmo"]["author"]["user_id"],
        author_name: v["data"]["gizmo"]["author"]["display_name"],
        created_at: v["created_at"],
        updated_at: v["data"]["gizmo"]["updated_at"],
        detail: JSON.stringify(v),
      });
    });

    return gpts;
  } catch (err) {
    console.error("Error loading JSON file:", err);
    return [];
  }
};

export const searchGpts = async (question: string): Promise<Gpts[]> => {
  const uri = `${process.env.INDEX_API_BASE_URI}/gpts/search`;
  const data = {
    question: question,
  };

  try {
    const resp = await fetch(uri, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${process.env.INDEX_API_KEY}`,
      },
      body: JSON.stringify(data),
    });
    const res = await resp.json();
    if (res.data) {
      return res.data.filter((gpts: Gpts) => !isGptsSensitive(gpts));
    }
  } catch (e) {
    console.log("request gpts search failed: ", e);
  }

  return [];
};

export function isGptsSensitive(gpts: Gpts): boolean {
  if (!gpts.name || !gpts.author_name || !gpts.description) {
    return true;
  }

  const sensitiveKeywords = process.env.SENSITIVE_KEYWORDS || "";
  const keywordsArr = sensitiveKeywords.split(",");
  for (let i = 0, l = keywordsArr.length; i < l; i++) {
    const keyword = keywordsArr[i].trim();
    if (
      gpts.name.includes(keyword) ||
      gpts.description.includes(keyword) ||
      gpts.author_name.includes(keyword)
    ) {
      console.log("gpt is sensitive: ", gpts.uuid, gpts.name, keyword);
      return true;
    }
  }

  return false;
}

export function gptGptsPromptStarters(gpts: Gpts): string[] | undefined {
  if (gpts.detail) {
    try {
      const v = gpts.detail;
      return v["data"]["gizmo"]["display"]["prompt_starters"];
    } catch (e) {
      console.log("parse gpts detail failed: ", e);
    }
  }

  return;
}

export function getGptsWelcomeMessage(gpts: Gpts): string | undefined {
  if (gpts.detail) {
    try {
      const v = gpts.detail;
      return v["data"]["gizmo"]["display"]["welcome_message"];
    } catch (e) {
      console.log("parse gpts detail failed: ", e);
    }
  }

  return;
}

export function getGptsTools(gpts: Gpts): string[] | undefined {
  if (gpts.detail) {
    try {
      const v = gpts.detail;
      let tools: string[] = [];
      v["data"]["tools"].forEach((tool: any) => {
        tools.push(tool["type"]);
      });
      return tools;
    } catch (e) {
      console.log("parse gpts detail failed: ", e);
    }
  }

  return;
}

File: index/models/gpts.py
import json
from pony.orm import db_session
from components.db import db


class Gpts:
    def __init__(self, v) -> None:
        self.uuid = v["data"]["gizmo"]["id"]
        self.org_id = v["data"]["gizmo"]["organization_id"]
        self.name = v["data"]["gizmo"]["display"]["name"]
        self.description = v["data"]["gizmo"]["display"]["description"]
        self.avatar_url = v["data"]["gizmo"]["display"]["profile_picture_url"]
        self.short_url = v["data"]["gizmo"]["short_url"]
        self.author_id = v["data"]["gizmo"]["author"]["user_id"]
        self.author_name = v["data"]["gizmo"]["author"]["display_name"]
        self.created_at = v["created_at"]
        self.updated_at = v["data"]["gizmo"]["updated_at"]


def get_gpts_from_file(file_name: str):
    with open(file_name, 'r', encoding='utf-8') as file:
        data = json.load(file)
        gpts = []
        for v in data:
            gpt = Gpts(v)
            print("gpts: ", gpt.name, gpt.uuid)
            gpts.append(gpt)
            
        return gpts


@db_session
def get_gpts_from_db(last_id: int, limit: int):
    if limit <= 0:
        limit = 100

    rows = db.select(
        "SELECT * FROM gpts WHERE id > $last_id AND index_updated_at = 0 ORDER BY id asc LIMIT $limit"
    )
    gpts = []
    for row in rows:
        gpts.append(row)

    return gpts


@db_session
def get_gpts_by_uuids(uuids: [str]):
    rows = db.select(
        "SELECT * FROM gpts WHERE uuid = ANY($uuids) ORDER BY id asc"
    )
    gpts = []
    for row in rows:
        gpts.append(row)

    return gpts


@db_session
def update_gpts_index_time(ids: [int], index_time: int):
    db.execute("UPDATE gpts SET index_updated_at=$index_time WHERE id = ANY($ids)")

    return



